core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -200,14 +229,32 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadTimestamps(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+    if (log.iterator.hasNext)
+      rollingBaseTimestamp = log.iterator.next.message.timestamp
+    val lastTimeIndexEntry = timeIndex.lastEntry
+    if (readToLogEnd) {
+      val offsetPosition = index.lookup(lastTimeIndexEntry.offset)
+      // Scan the rest of the messages to see if there is a larger timestamp after the last time index entry.
+      maxTimestampSoFar = math.max(lastTimeIndexEntry.timestamp, log.largestTimestampAfter(offsetPosition.position))
+    } else
+      maxTimestampSoFar = lastTimeIndexEntry.timestamp
```

Comment:
If we update maxTimestampSoFar, don't we need to update offsetOfMaxTimestamp accordingly? Otherwise, the next time we add an entry to the time index, the offset may not match maxTimestampSoFar.